### PR TITLE
Insert BLOB from bytes or memoryview. Copy BLOB as is to bytes object

### PR DESCRIPTION
**Pull request by penguian**
Addresses issue #1703. 

I added `Blob::ToStringAsIs`, `Value::ToStringAsIs` and `Value::GetValueAsIs`, and changed `DuckDBPyResult::Fetchone()` only for `LogicalTypeId::BLOB`, to avoid changing existing behaviour. The names `ToStringAsIs` and `GetValueAsIs` are suggestions. You can change these to whatever makes more sense to you, flesh them out so that they treat types other than `Blob`, etc.

Actually, I don't understand why `Value::GetValue` on a `BLOB` type calls `Blob::ToString` which transforms non-printing ASCII chars to `\xnn` hex escape sequences. Why is `BLOB` being transformed into a printable ASCII string when it should just be passed back as-is? Should I create another issue to explore this question in more detail?

For example, `test_CheckMemoryviewFromhexBlob` checks that a `bytes` object containing null (hex '00') characters can be inserted into a `BLOB` column and fetched back unchanged. Without the change to `DuckDBPyResult::Fetchone()` the fetch returns a printable string containing (e.g.) `\x00` sequences.

### CODE REVIEW

**Review by hawkfish — COMMENTED** [review]
Thanks for the PR! The code looks fine, but I agree we should think about the name a bit. `ToStringRaw`? `ToBytes`?

File: src/common/types/value.cpp [hunk 1]
```diff
@@ -927,6 +931,18 @@ string Value::ToString() const {
 	}
 }
 
+string Value::ToStringAsIs() const {
+	if (is_null) {
+		return "NULL";
+	}
+	switch (type_.id()) {
+	case LogicalTypeId::BLOB:
+		return Blob::ToStringAsIs(string_t(str_value));
```
Comment (Mytherin) [inline src/common/types/value.cpp:940]: We can just directly return `str_value` here, no? That should simplify the code here.

File: src/common/types/blob.cpp [hunk 2]
```diff
@@ -63,6 +63,20 @@ string Blob::ToString(string_t blob) {
 	return string(buffer.get(), str_len);
 }
 
+void Blob::ToStringAsIs(string_t blob, char *output) {
```
Comment (Mytherin) [inline src/common/types/blob.cpp:66]: We also won't need these two functions anymore then, right?

File: src/include/duckdb/common/types/value.hpp [hunk 3]
```diff
@@ -124,6 +124,10 @@ class Value {
 		throw NotImplementedException("Unimplemented template type for Value::GetValue");
 	}
 	template <class T>
+	T GetValueAsIs() const {
```
Comment (Mytherin) [inline src/include/duckdb/common/types/value.hpp:127]: Rather than creating an additional template function that only works for blobs, we could just directly use `ToStringAsIs`, no? That should clean up the code here as well somewhat.

File: src/include/duckdb/common/types/value.hpp [hunk 4]
```diff
@@ -139,6 +143,9 @@ class Value {
 		return Value(*this);
 	}
 
+	//! Convert this value to a string as is
+	DUCKDB_API string ToStringAsIs() const;
```
Comment (Mytherin) [inline src/include/duckdb/common/types/value.hpp:147]: Agreed with Richard, perhaps we can call this `GetBlob` to be consistent with the `GetValue` syntax, and our definition of the type as blob?

File: tools/pythonpkg/tests/sqlite/test_types.py [hunk 5]
```diff
@@ -81,6 +81,28 @@ def test_CheckInf(self):
...
+        val = memoryview(sample)
+        self.cur.execute("insert into test(b) values (?)", (val,))
+        self.cur.execute("select b from test")
+        row = self.cur.fetchone()
+        self.assertEqual(row[0], sample)
+
+    def test_CheckMemoryviewFromhexBlob(self):
+        sample = bytes.fromhex('00FF0F2E3D4C5B6A798800FF00')
+        val = memoryview(sample)
+        self.cur.execute("insert into test(b) values (?)", (val,))
```
Comment (Mytherin) [inline tools/pythonpkg/tests/sqlite/test_types.py:102]: Could you add a test that passes `None` as a template parameter to a blob?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good already, great that you got it to work! Some comments that could help simplify the code:

**Comment by Mytherin** [comment]
> Actually, I don't understand why Value::GetValue on a BLOB type calls Blob::ToString which transforms non-printing ASCII chars to \xnn hex escape sequences. Why is BLOB being transformed into a printable ASCII string when it should just be passed back as-is? Should I create another issue to explore this question in more detail?

The problem there is template overloading, when you specify `GetValue<string>` we can't distinguish between whether you want a blob or a string type, so we assume you want a string type as that is the most common case. We then call `ToString` to perform the conversion, which returns a stringified representation of the type. That behavior seems fine to me, but indeed an extra function to expose the blob data itself is required.

File: src/common/types/value.cpp (hunk 1 above)
Comment (penguian) [reply src/common/types/value.cpp]: I reverted this change.

File: src/common/types/blob.cpp (hunk 2 above)
Comment (penguian) [reply src/common/types/blob.cpp]: I reverted this change.

File: src/include/duckdb/common/types/value.hpp (hunk 4 above)
Comment (penguian) [reply src/include/duckdb/common/types/value.hpp]: I reverted this change.

File: tools/pythonpkg/tests/sqlite/test_types.py (hunk 5 above)
Comment (penguian) [reply tools/pythonpkg/tests/sqlite/test_types.py:102]: I need to make this test dependent on Python 3, since Python 2 does not have `fromhex()`.

Comment (Mytherin) [reply tools/pythonpkg/tests/sqlite/test_types.py:102]: Sure, that's fine.

File: src/common/types/value.cpp (hunk 1 above)
Comment (Mytherin) [reply src/common/types/value.cpp]: Looks great!

**Comment by Mytherin** [comment]
The format fix test still seems to fail, could you run `make format-fix` to fix this? Alternatively you can also manually fix the issues reported in the CI run.

File: tools/pythonpkg/src/pyconnection.cpp [hunk 6]
```diff
@@ -502,6 +500,15 @@ vector<Value> DuckDBPyConnection::TransformPythonParamList(py::handle params) {
...
 			args.push_back(Value::DATE(year, month, day));
+		} else if (py::isinstance<py::str>(ele)) {
+			args.emplace_back(ele.cast<string>());
+		} else if (py::isinstance<py::memoryview>(ele)) {
+			py::memoryview py_view = ele.cast<py::memoryview>();
+			PyObject* py_view_ptr = py_view.ptr();
+			Py_buffer* py_buf = PyMemoryView_GET_BUFFER(py_view_ptr);
+			args.emplace_back(Value::BLOB(const_data_ptr_t(py_buf->buf), idx_t(py_buf->len)));
+		} else if (py::isinstance<py::bytes>(ele)) {
+			args.emplace_back(Value::BLOB(ele.cast<string>()));
```
Comment (penguian) [inline tools/pythonpkg/src/pyconnection.cpp:511]: Creating a `BLOB` from a `string` currently translates `\xnn` sequences to byte values. Correct? `TransformPythonParamList` here needs to just store the string without translation.

Comment (Mytherin) [reply tools/pythonpkg/src/pyconnection.cpp:511]: Yes, in that case you can use the alternative blob interface:
```cpp
//! Create a blob Value from a data pointer and a length: no bytes are interpreted
static Value BLOB(const_data_ptr_t data, idx_t len);
//! Creates a blob by casting a specified string to a blob (i.e. interpreting \x characters)
static Value BLOB(const string &data);
```

**Review by hawkfish — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks!
